File: backend/reports/evidence.py
# backend/reports/evidence.py
import uuid
from datetime import datetime, timezone


def build_evidence(account_id: str, scores_map: dict, G) -> dict:
    sc = scores_map.get(account_id, {})

    node_attrs = G.nodes.get(account_id, {})
    neighbors = set(G.predecessors(account_id)) | set(G.successors(account_id))

    timeline = []
    for u, v, attrs in G.edges(data=True):
        if u == account_id or v == account_id:
            timeline.append({
                "from": u, "to": v,
                "amount": attrs.get("amount", 0),
                "type": attrs.get("type", ""),
                "timestamp": str(attrs.get("timestamp", "")),
                "is_fraud": attrs.get("is_fraud", 0),
            })
    timeline.sort(key=lambda x: x["timestamp"])

    return {
        "report_id": str(uuid.uuid4()),
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "subject_account": account_id,
        "risk_score": sc.get("risk_score", 0),
        "triggered_patterns": sc.get("triggered_patterns", []),
        "transaction_timeline": timeline,
        "connected_accounts": list(neighbors),
        "shap_breakdown": sc.get("shap_breakdown", {}),
        "llm_narrative": "",
        "investigator_notes": "",
    }

File: frontend/lib/models/evidence.dart
class Evidence {
  final String reportId;
  final String generatedAt;
  final String subjectAccount;
  final double riskScore;
  final List<String> triggeredPatterns;
  final List<Map<String, dynamic>> transactionTimeline;
  final List<String> connectedAccounts;
  final Map<String, double> shapBreakdown;
  final String llmNarrative;
  final String investigatorNotes;

  const Evidence({
    required this.reportId,
    required this.generatedAt,
    required this.subjectAccount,
    required this.riskScore,
    required this.triggeredPatterns,
    required this.transactionTimeline,
    required this.connectedAccounts,
    required this.shapBreakdown,
    required this.llmNarrative,
    required this.investigatorNotes,
  });

  factory Evidence.fromJson(Map<String, dynamic> j) => Evidence(
        reportId: j['report_id'] as String,
        generatedAt: j['generated_at'] as String,
        subjectAccount: j['subject_account'] as String,
        riskScore: (j['risk_score'] as num).toDouble(),
        triggeredPatterns: (j['triggered_patterns'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
        transactionTimeline:
            (j['transaction_timeline'] as List<dynamic>? ?? [])
                .map((e) => e as Map<String, dynamic>)
                .toList(),
        connectedAccounts: (j['connected_accounts'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
        shapBreakdown: (j['shap_breakdown'] as Map<String, dynamic>? ?? {})
            .map((k, v) => MapEntry(k, (v as num).toDouble())),
        llmNarrative: j['llm_narrative'] as String? ?? '',
        investigatorNotes: j['investigator_notes'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {
        'report_id': reportId,
        'generated_at': generatedAt,
        'subject_account': subjectAccount,
        'risk_score': riskScore,
        'triggered_patterns': triggeredPatterns,
        'transaction_timeline': transactionTimeline,
        'connected_accounts': connectedAccounts,
        'shap_breakdown': shapBreakdown,
        'llm_narrative': llmNarrative,
        'investigator_notes': investigatorNotes,
      };
}